File: java-kata/src/main/java/com/leetcode/RansomNote.java
package com.leetcode;

/**
 * <h3><a href="https://leetcode.com/problems/ransom-note/">Ransom Note</a></h3>
 * <p>
 *     Given two strings <code>ransomNote</code> and <code>magazine</code>, return <code>true</code>
 *     if <code>ransomNote</code> can be constructed by using the letters from <code>magazine</code>
 *     and <code>false</code> otherwise.
 * </p>
 * <p>
 *     Each letter in <code>magazine</code> can only be used once in <code>ransomNote</code>.
 * </p>
 * <p>
 *     <h6>Constraints:</h6>
 *     <ul>
 *         <li><code>1 <= ransomNote.length, magazine.length <= 10^5</code></li>
 *         <li><code>ransomNote</code> and <code>magazine</code> consist of lowercase English letters</li>
 *     </ul>
 * </p>
 */
public class RansomNote {

    public boolean canConstruct(String ransomNote, String magazine) {
        final int[] freqs = new int[26];
        for (int i = 0; i < magazine.length(); i++) {
            freqs[magazine.charAt(i) - 'a']++;
        }
        for (int i = 0; i < ransomNote.length(); i++) {
            if (freqs[ransomNote.charAt(i) - 'a'] == 0) {
                return false;
            }
            freqs[ransomNote.charAt(i) - 'a']--;
        }
        return true;
    }
}

File: kotlin-kata/src/main/kotlin/com/leetcode/RansomNote.kt
package com.leetcode

/**
 * ### [Ransom Note](https://leetcode.com/problems/ransom-note/)
 *
 * Given two strings `ransomNote` and `magazine`, return `true`
 * if `ransomNote` can be constructed by using the letters from `magazine`
 * and `false` otherwise.
 *
 * Each letter in `magazine` can only be used once in `ransomNote`.
 *
 * ###### Constraints:
 *  - `1 <= ransomNote.length, magazine.length <= 10^5`
 *  - `ransomNote` and `magazine` consist of lowercase English letters
 */
class RansomNote {

    fun canConstruct(ransomNote: String, magazine: String): Boolean {
        var freqs = IntArray(26)

        magazine.forEach {
            freqs[it - 'a']++
        }

        ransomNote.forEach {
            if (freqs[it - 'a'] == 0) {
                return false
            }
            freqs[it - 'a']--
        }

        return true
    }
}
